keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+ }
+```
+
+##### Default Endpoint IP Selection
+Older servers and clients that were built before the introduction of full dual stack will only be aware of and make use of the original, singular IP field above. It is therefore considered to be the default IP address for the endpoint. When the IP and IPs fields are populated, the IPs[0] field must match the (default) IP entry. If a pod has both IPv4 and IPv6 addresses allocated, then the IP address chosen as the default IP address will match the IP family of the cluster's configured service CIDR. For example, if the service CIDR is IPv4, then the IPv4 address will be used as the default endpoint address.
+
+#### EndpointAddress Internal Representation
+The EndpointAddress internal representation will be modified to use a slice of IP strings ("IPs") rather than a singular IP strings ("IP"):
+```
+// EndpointAddress is a tuple that describes one or more IP addresses for an
+// endpoint.
+type EndpointAddress struct {
+	// The IPs for this endpoint.
+	IPs []string
+	// Optional: Hostname of this endpoint
+	// Meant to be used by DNS servers etc.
+	// +optional
+	Hostname string
+	// Optional: Node hosting this endpoint. This can be used to determine endpoints local to a node.
+	// +optional
+	NodeName *string
+	// Optional: The kubernetes object related to the entry point.
+	TargetRef *ObjectReference
+}
+```
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "IP" field).
+
+#### Maintaining Compatible Interworking between Old and New Clients
+With this API change, we need to consider the scenario where there are a mix of clients that are running old vs. new versions of the API. The old clients would only be aware of the original, singular IP field, whereas newer clients could be writing either just the plural IPs field, or updating both singular IP and plural IPs fields. To cover this case, the API server will need to implement some safeguards/fix-ups for these fields for write operations (in compliance with the guidelines for pluralizing singular API fields that is included in the [Kubernetes API change quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md):
+
+##### V1 to Core (Internal) Conversion
+- If only V1 IP is provided:
+  - Copy V1 IP to core IPs[0]
+- Else if only V1 IPs[] is provided:
+  - Copy V1 IPs[] to core IPs[]
+- Else if both V1 IP and V1 IPs[] are provided:
+  - Verify that V1 IP matches V1 IPs[0]
+  - Copy V1 IPs[] to core IPs[]
+- Delete any duplicates in core IPs[]
+
+##### Core (Internal) to V1 Conversion
+  - Copy core IPs[0] to V1 IP
+  - Copy core IPs[] to V1 IPs[]
+
+
+#### Configuration of Endpoint IP Family in Service Definitions
+This proposal adds an option to configure an endpoint IP family for a Kubernetes service:
+```
+    endpointFamily: <ipv4|ipv6|dual-stack>       [Default: dual-stack]
```

Comment:
If service addresses only come from a single address family, why does this belong in the Service definition?

Or to put it another way - shouldn't the default be the same address family as the service CIDR? If Kubernetes itself isn't going to do any 6/4 translation, could you say more about how this can be used in any other way?